dart-impl/base/src/internal/domain_locality.c

```diff
@@ -490,10 +590,98 @@ dart_ret_t dart__base__locality__domain__filter_subdomains(
   return DART_OK;
 }
 
+dart_ret_t dart__base__locality__domain__add_subdomain(
+  dart_domain_locality_t           * domain,
+  dart_domain_locality_t           * subdomain,
+  int                                subdomain_rel_id)
+{
+  /* inserts pointer to subdomain and updates number of subdomains but
+   * does not recalculate domain attributes.
+   */
+  domain->num_domains++;
+  dart_domain_locality_t ** tmp;
+  tmp = realloc(domain->children,
+                  domain->num_domains *
+                    sizeof(dart_domain_locality_t *));
+
+  if (NULL != tmp) {
+    domain->children = tmp;
+  } else {
+      DART_ASSERT_MSG(
+        0, "dart__base__locality__domain__add_subdomain: "
+           "realloc failed");
+      return DART_ERR_OTHER;
+  }
+  if (subdomain_rel_id < 0) {
+    /* append at end of subdomains: */
+    domain->children[domain->num_domains - 1] = subdomain;
+  } else {
+    /* move all subdomains after insertion index: */
+    for (int sd_move = subdomain_rel_id;
+         sd_move < domain->num_domains-1; sd_move++) {
+      domain->children[sd_move + 1] = domain->children[sd_move];
+    }
+    domain->children[subdomain_rel_id] = subdomain;
+  }
+  return DART_OK;
+}
+
+dart_ret_t dart__base__locality__domain__remove_subdomain(
+  dart_domain_locality_t           * domain,
+  int                                subdomain_rel_id)
+{
+  // TODO: also remove the subdomain from all aliases lists
+
+  /* reduces number of subdomains but does not recalculate domain
+   * attributes and does not destruct the removed subdomain.
+   */
+
+  /* move all subdomains after removed index: */
+  for (int sd_move = subdomain_rel_id;
+       sd_move < domain->num_domains-1; sd_move++) {
+    domain->children[sd_move] = domain->children[sd_move + 1];
+  }
+  domain->num_domains--;
+
+  if (domain->num_domains <= 0) {
+    domain->children    = NULL;
```

Comment:
There is no memory to leak, `domain->children` is an array of pointers and will be realloced if needed.
The removed domain object of type `dart_domain_locality_t` is not deleted in this function, just "unlinked" from the specified parent domain.